clippy_lints/src/types.rs

```diff
@@ -1558,15 +1584,55 @@ fn lint_fn_to_numeric_cast(
 ) {
     // We only want to check casts to `ty::Uint` or `ty::Int`
     match cast_to.kind {
-        ty::Uint(_) | ty::Int(..) => { /* continue on */ },
+        ty::Uint(_) | ty::Int(_) => { /* continue on */ },
         _ => return,
     }
-    match cast_from.kind {
-        ty::FnDef(..) | ty::FnPtr(_) => {
-            let mut applicability = Applicability::MaybeIncorrect;
-            let from_snippet = snippet_with_applicability(cx, cast_expr.span, "x", &mut applicability);
 
-            let to_nbits = int_ty_to_nbits(cast_to, cx.tcx);
+    let is_def = matches!(cast_from.kind, ty::FnDef(..));
+    let is_ptr = matches!(cast_from.kind, ty::FnPtr(_));
+
+    if is_def || is_ptr {
+        let mut applicability = Applicability::MaybeIncorrect;
+        let from_snippet = snippet_with_applicability(cx, cast_expr.span, "x", &mut applicability);
+
+        let to_nbits = int_ty_to_nbits(cast_to, cx.tcx);
+
+        if is_def {
+            if to_nbits < cx.tcx.data_layout.pointer_size.bits() {
+                span_lint_and_sugg(
+                    cx,
+                    FN_TO_NUMERIC_CAST_WITH_TRUNCATION,
+                    expr.span,
+                    &format!(
+                        "casting *pointer* to function `{}` as `{}`, which truncates the value",
+                        from_snippet, cast_to
+                    ),
+                    "try",
+                    format!("{}() as {}", from_snippet, cast_to),
+                    applicability,
+                );
+            } else if cast_to.kind == ty::Uint(UintTy::Usize) {
+                span_lint_and_sugg(
+                    cx,
+                    FN_TO_NUMERIC_CAST_USIZE,
+                    expr.span,
+                    &format!("casting *pointer* to function `{}` as `{}`", from_snippet, cast_to),
+                    "try",
+                    format!("{}() as {}", from_snippet, cast_to),
+                    applicability,
+                );
+            } else {
+                span_lint_and_sugg(
+                    cx,
+                    FN_TO_NUMERIC_CAST,
+                    expr.span,
+                    &format!("casting *pointer* to function `{}` as `{}`", from_snippet, cast_to),
+                    "try",
+                    format!("{}() as {}", from_snippet, cast_to),
```

Comment:
I don't think we should assume here, that a cast of the return value was meant. Also I think the `casting function pointer ...` message was clear enough that it talks about a function pointer. Also emphasizing with * or something else isn't used in error messages.

What we can do is add two suggestions, one that suggests to cast the pointer to `usize` and one to call the function and cast the resulting value to `usize`, iff it is a `FnDef(..)`